TDEAL-27: Fix Content Security Policy with Browser Link

sarahelsaig opened the pull request:
[TDEAL-27](https://lombiq.atlassian.net/browse/TDEAL-27)

[TDEAL-27]: https://lombiq.atlassian.net/browse/TDEAL-27?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sarahelsaig committed:
Exclude admin.

sarahelsaig committed:
SuppressMessage

sarahelsaig committed:
Also permit jquery from CDN.

sarahelsaig committed:
Add SkipContentSecurityPolicyProvider.

sarahelsaig committed:
Fix header.

sarahelsaig committed:
fix analyzer violations

sarahelsaig committed:
Add BrowserLinkContentSecurityPolicyProvider.

sarahelsaig committed:
unusing

sarahelsaig committed:
Use correct context.

sarahelsaig committed:
Fix directive.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/TDEAL-27-main

# Conflicts:
#	Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs

sarahelsaig committed:
Fix warnings.

Piedone commented on Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs:
```diff
@@ -32,6 +32,7 @@ public class CdnContentSecurityPolicyProvider : IContentSecurityPolicyProvider
     public static ConcurrentBag<Uri> PermittedScriptSources { get; } = new(new[]
     {
         new Uri("https://cdn.jsdelivr.net/npm"),
+        new Uri("https://code.jquery.com/jquery-3.7.0.js"),
```
Why include just this one JS and not the whole of https://code.jquery.com? We'd need to keep this up-to-date with OC versions (and the current one there is actually 3.7.1).

sarahelsaig replied:
Actually [only the host names](https://github.com/Lombiq/Helpful-Libraries/blob/09defab4fe25912372bf3a12e85b6136f3790f78/Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs#L85) of these URLs matter. So this is the same as
```suggestion
        new Uri("https://code.jquery.com/"),
```
I can remove the `jquery-3.7.0.js` if you think it's confusing.

Piedone replied:
It is, yes, please.

Piedone commented on Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs:
```diff
@@ -32,6 +32,7 @@ public class CdnContentSecurityPolicyProvider : IContentSecurityPolicyProvider
     public static ConcurrentBag<Uri> PermittedScriptSources { get; } = new(new[]
```
While we're at it, let's prevent future such issues by adding all default CDNs used by OC. `https://maxcdn.bootstrapcdn.com` is one too, but check out `OrchardCore.Resources`.

sarahelsaig replied:
These are the unique hosts used in [`OrchardCore.Resources.ResourceManagementOptionsConfiguration`](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Resources/ResourceManagementOptionsConfiguration.cs):
- https://cdn.jsdelivr.net/
- https://cdnjs.cloudflare.com/
- https://code.jquery.com/
- https://maxcdn.bootstrapcdn.com/

Added the two highlighted that were missing.

Piedone committed:
Code styling

Piedone committed:
Code styling

Piedone commented on Lombiq.HelpfulLibraries.OrchardCore/Security/SkipContentSecurityPolicyProvider.cs:
```diff
@@ -0,0 +1,30 @@
+    private bool ShouldSuppressHeaderInner(HttpContext context) =>
+        // No need to do content security policy on non-HTML responses.
+        context.Response.ContentType?.ContainsOrdinalIgnoreCase(MediaTypeNames.Text.Html) != true ||
+        // The Admin dashboard is only accessible to trusted users so content security is not a concern.
+        new Uri(context.Request.GetDisplayUrl()).AbsolutePath.StartsWithOrdinalIgnoreCase(_adminPathPrefix);
```
It can be a concern since you can do an XSS attack from the frontend, that'll be only active on the admin (like loading a script in a WYSIWYG editor). The admin should be guarded too.

sarahelsaig replied:
Well okay, but we have to permit eval due to the media field editor using Vue anyway so I'm a bit dubious how helpful that is.

Piedone replied:
That can be allowed only under specific URLs.

sarahelsaig replied:
What do you mean? The URL would be the same `/Admin/Contents/ContentItems/{GUID}/Edit` where a WYSIWYG editor might appear.

Piedone replied:
Yeah, that one. Doesn't the Media Library need that too, BTW?

sarahelsaig replied:
Yep. Did that in TDEAL

Piedone replied:
Why not here too?

sarahelsaig replied:
Because it's kludge.

Piedone replied:
Can you elaborate? Won't we have exactly the same issue in all of our projects using Media and the HL security default?

sarahelsaig replied:
I'm not expecting that, considering that we have yet to find a whole second person who can manifest the problem in question. I works flawlessly for you, me and PG, nobody else have yet complained about it in OCC, it was never a problem in UI tests either. I'd rather have more data points.

sarahelsaig committed:
Add domains from OrchardCore.Resources to CdnContentSecurityPolicyProvider.

sarahelsaig committed:
Add domains from OrchardCore.Resources to CdnContentSecurityPolicyProvider.

# Conflicts:
#	Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs

sarahelsaig committed:
Merge branch 'issue/TDEAL-27-main' of https://github.com/Lombiq/Helpful-Libraries into issue/TDEAL-27-main

sarahelsaig committed:
Don't skip Admin.

sarahelsaig committed:
Merge tag 'v8.1.1' into issue/TDEAL-27

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/TDEAL-27' into issue/TDEAL-27-main

# Conflicts:
#	Lombiq.HelpfulLibraries.AspNetCore/Security/CdnContentSecurityPolicyProvider.cs

sarahelsaig committed:
Code styling.

Piedone committed:
Merge remote-tracking branch 'origin/dev' into issue/TDEAL-27-main
